lib/openssl/buffering.rb

```diff
@@ -345,13 +345,18 @@ def do_write(s)
     @wbuffer << s
     @wbuffer.force_encoding(Encoding::BINARY)
     @sync ||= false
-    if @sync or @wbuffer.size > BLOCK_SIZE
-      until @wbuffer.empty?
-        begin
-          nwrote = syswrite(@wbuffer)
-        rescue Errno::EAGAIN
-          retry
+    buffer_size = @wbuffer.size
+    if @sync or buffer_size > BLOCK_SIZE
+      nwrote = 0
+      begin
+        while nwrote < buffer_size do
+          begin
+            nwrote += syswrite(@wbuffer[nwrote..buffer_size])
```

Comment:
```suggestion
            nwrote += syswrite(@wbuffer[nwrote, buffer_size - nwrote])
```

can save one Range object allocation in the loop.

Reply:
Oh hey, this is totally better - updated!